substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.jfr.traceid;
+
+import com.oracle.svm.core.annotate.Uninterruptible;
+import com.oracle.svm.core.heap.Heap;
+import com.oracle.svm.core.thread.VMOperation;
+
+import java.util.function.Consumer;
+
+public class JfrTraceIdLoadBarrier {
+    private static int classCount0;
+    private static int classCount1;
+
+    @Uninterruptible(reason = "Epoch may not change")
+    private static boolean isNotTagged(long value) {
+        long thisEpochBit = JfrTraceIdEpoch.getInstance().thisEpochBit();
+        return ((value & ((thisEpochBit << JfrTraceId.META_SHIFT) | thisEpochBit)) != thisEpochBit);
+    }
+
+    @Uninterruptible(reason = "Epoch may not change")
+    private static boolean shouldTag(Class<?> obj) {
+        assert obj != null;
+        return isNotTagged(JfrTraceId.getTraceIdRaw(obj));
+    }
+
+    public static void clear() {
+        clearClassCount(JfrTraceIdEpoch.getInstance().previousEpoch());
```

Comment:
I suppose this is not marked with `@Uninterruptible` because will only be executed at a safepoint? If so, please add an assertion.